kuzminskiy_stafeev/graph_generator.cpp

```diff
@@ -74,41 +102,121 @@ void generate_red_edges(Graph& graph, Graph::Depth depth) {
+
+  auto vertex_id = graph.add_vertex();
+  graph.add_edge(from_vertex_id, vertex_id);
+  cur_vertex_ids.emplace_back(vertex_id);
+}
+
 }  // namespace
 
 Graph GraphGenerator::generate() const {
   auto graph = Graph();
   if (params_.depth()) {
     graph.add_vertex();
     generate_grey_edges(graph);
-    generate_green_edges(graph);
-    generate_yellow_edges(graph, graph.get_graph_depth());
-    generate_red_edges(graph, graph.get_graph_depth());
+    std::thread green_thread(generate_green_edges, std::ref(graph));
+    std::thread yellow_thread(&generate_yellow_edges, std::ref(graph),
+                              graph.get_graph_depth());
+    std::thread red_thread(&generate_red_edges, std::ref(graph),
+                           graph.get_graph_depth());
+
+    green_thread.join();
+    yellow_thread.join();
+    red_thread.join();
   }
+
   return graph;
 }
 
+void GraphGenerator::generate_grey_branch(
+    Graph& graph,
+    std::mutex& graph_mutex,
+    std::vector<Graph::VertexId> previous_vertex_ids,
+    Graph::Depth cur_depth) const {
+  const auto depth = params_.depth();
+  std::vector<Graph::VertexId> cur_vertex_ids;
+
+  if (depth < cur_depth) {
+    return;
+  } else if (cur_depth == 2) {
+    add_grey_edge(graph, 0, graph_mutex, cur_vertex_ids);
+    generate_grey_branch(graph, graph_mutex, cur_vertex_ids, cur_depth + 1);
+    return;
+  }
+
+  float step = 1.0 / (depth - 1);
+  float prob = (float)(1 - step * (cur_depth - 2));
+
+  const auto new_vertices_count = params_.new_vertices_count();
+
+  for (const auto from_vertex_id : previous_vertex_ids) {
+    for (Graph::VertexId cur_vertex_count = 0;
+         cur_vertex_count < new_vertices_count; cur_vertex_count++) {
+      if (check_probability(prob)) {
+        add_grey_edge(graph, from_vertex_id, graph_mutex, cur_vertex_ids);
+      }
+    }
+  }
+
+  generate_grey_branch(graph, graph_mutex, cur_vertex_ids, cur_depth + 1);
+}
+
 void GraphGenerator::generate_grey_edges(Graph& graph) const {
   const auto depth = params_.depth();
+  if (depth == 1) {
+    return;
+  }
+
+  auto jobs = std::list<JobCallBack>();
   const auto new_vertices_count = params_.new_vertices_count();
-  float step = 1.0 / (depth - 1);
+  std::mutex graph_mutex;
+  std::vector<Graph::VertexId> base_vertex_id = {0};
 
-  for (Graph::Depth cur_depth = 1; cur_depth <= depth; cur_depth++) {
-    float prob = (float)(1 - step * (cur_depth - 1));
-    bool is_created = false;
-    for (const auto from_vertex_id : graph.vertices_of_depth(cur_depth)) {
-      for (Graph::VertexId cur_vertex_count = 0;
-           cur_vertex_count < new_vertices_count; cur_vertex_count++) {
-        if (check_probability(prob)) {
-          is_created = true;
-          auto vertex_id = graph.add_vertex();
-          graph.add_edge(from_vertex_id, vertex_id);
+  for (int i = 0; i < new_vertices_count; i++) {
+    jobs.push_back(
+        [&]() { generate_grey_branch(graph, graph_mutex, base_vertex_id, 2); });
```

Comment:
1) Считается плохой практикой захватывать всё, нужно перечислять все переменные, которые нужны в лямбде
    ```suggestion
            [&graph, &graph_mutex, &base_vertex_id, this]() { generate_grey_branch(graph, graph_mutex, base_vertex_id, 2); });
    ```
    Захват `this` нужен для вызова `generate_grey_branch()`, так как это член класса, то под капотом он вызывается так: `this->generate_grey_branch()`, поэтому нам и нужно захватить `this`
2) 2 - `magic number`